### PR TITLE
Add docs on how to fix flaky tests

### PR DESCRIPTION
**Pull request by JelteF**
I fixed a lot of flaky tests recently and I found some patterns in the
type of issues and type of fixes. This adds a document that lists 
these types of issues and explains how to fix them.

### CODE REVIEW

File: src/test/regress/flaky_tests.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+
+**Issue**: The logs in the regress output are displayed in a different order
+than what the output file shows
+
+**Fix**: It's simple: don't log these things during the test. There's two common
+ways of achieving this:
+1. If you don't care about the logs for this query at all, then you can change
+   the log `VERBOSITY`
+2. If this is are logs created by `citus.log_remote_commands` and you care about
```
Comment (gledis69) [inline src/test/regress/flaky_tests.md:53]:
```suggestion
2. If these are logs created by `citus.log_remote_commands` and you care about
```

File: src/test/regress/flaky_tests.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+   some of the remote commands being as expected, but not the ones in question.
+   Then you can use `citus.grep_remote_commands` to only display the commands
```
Comment (gledis69) [inline src/test/regress/flaky_tests.md:55]: NIT: up to you
```suggestion
   some of the remote commands being as expected, 
   then you can use `citus.grep_remote_commands` to only display the commands
```

File: src/test/regress/flaky_tests.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+   that you care about.
+
+
+
+**Example 1**:
```
Comment (gledis69) [inline src/test/regress/flaky_tests.md:60]: Very NIT: Can you make the Example/Fix pair consistent across the file.
e.g. example 1/fix 1; example of issue 1/example of fix 1

File: src/test/regress/flaky_tests.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+```
+
+**Example of fix 2**
+```diff
+ SET citus.log_remote_commands TO ON;
++SET citus.grep_remote_commands = '%ANALYZE%';
+```
+
+
+### Isolation test completes in different order
```
Comment (gledis69) [inline src/test/regress/flaky_tests.md:112]: Some isolation tests steps will also flakily wait sometimes due to step duration, and there is a special syntax to always make the test wait regardless of how long it takes. I remember @hanefi used in a flaky test fix before. You or Hanefi might want to add that here as well.

File: src/test/regress/flaky_tests.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+Citus its adaptive executor sometimes opens extra connections to do stuff in
+parallel. This happens especially in CI, because CI machines are sometimes slow.
+To disable parallelism and get consistent you can use `citus.max_adaptive_executor_pool_size`.
+
+**Example**
+```
++SET citus.max_adaptive_executor_pool_size TO 1;
+ ALTER TABLE dist_partitioned_table ADD CONSTRAINT constraint1 UNIQUE (dist_col, partition_col);
++RESET citus.max_adaptive_executor_pool_size TO 1;
+```
```
Comment (gledis69) [inline src/test/regress/flaky_tests.md:205]: Can you also do an example/fix pair here.

File: src/test/regress/flaky_tests.md [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+**Example of fix 2**
+```diff
+ SET citus.log_remote_commands TO ON;
++SET citus.grep_remote_commands = '%ANALYZE%';
+```
+
+
+### Isolation test completes in different order
+
+**Issue**: The two sessions in the isolation test get unblocked at the same
```
Comment (onderkalaci) [inline src/test/regress/flaky_tests.md:114]: pretty minor:

I find this wording a little confusing for `Issue` description; `get unblocked at the same time` 

I think the issue is more like: `there is no strict ordering between steps as there is not explicit locking between them`.   

And then the outcome of the issue is what you write. (so, happy to have (variations of) both sentences)

File: src/test/regress/flaky_tests.md [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+## What to do if this all doesn't work?
+
+If none of the advice above worked, the first thing to try is read the failing
+test in detail and try to understand how it works. Often, with a bit of thinking
+you can figure out why it's failing in the way that it's failing. If you cannot
+figure it out yourself, it's good to ask senior/principal engineers, maybe they
+can think of the reason. Or maybe they're certain that it's an actual bug.
+
+
+## What to do if output can never be consistent?
```
Comment (onderkalaci) [inline src/test/regress/flaky_tests.md:220]: would be super useful to add one example to this section as well. I have seen many people struggling to add a good one (including myself;) ))

File: src/test/regress/flaky_tests.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+Citus its adaptive executor sometimes opens extra connections to do stuff in
+parallel. This happens especially in CI, because CI machines are sometimes slow.
+To disable parallelism and get consistent you can use `citus.max_adaptive_executor_pool_size`.
+
+**Example**
+```
++SET citus.max_adaptive_executor_pool_size TO 1;
+ ALTER TABLE dist_partitioned_table ADD CONSTRAINT constraint1 UNIQUE (dist_col, partition_col);
++RESET citus.max_adaptive_executor_pool_size TO 1;
```
Comment (onderkalaci) [inline src/test/regress/flaky_tests.md:204]: I guess the right syntax would be `RESET citus.max_adaptive_executor_pool_size`

**Review by onderkalaci — APPROVED** [review]
Very interesting (and useful) PR! 

Can we put any references of this file in https://github.com/citusdata/citus/blob/main/CONTRIBUTING.md?

I have some minor comments. I'm approving it but I'd probably wait for others to comment as well before merging, given everyone might have different ideas on improving this readme.

File: src/test/regress/flaky_tests.md [hunk 9]
```diff
@@ -0,0 +1,227 @@
+# How to fix flaky tests
+
+There's three different causes of flaky tests.
```
Comment (naisila) [inline src/test/regress/flaky_tests.md:3]: Grammar: There are ...

File: src/test/regress/flaky_tests.md [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+1. Tests that don't make sure output is consistent, i.e. a bug in our tests
+2. Bugs in our testing infrastructure
+3. Bugs in Citus itself
+
+All of these impact the happiness of our developers, because we have to rerun
+tests to make them pass. But apart for developer happiness 3 also impacts our
```
Comment (naisila) [inline src/test/regress/flaky_tests.md:9]: apart for -> apart from

File: src/test/regress/flaky_tests.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+There are three different causes of flaky tests:
+1. Tests that don't make sure output is consistent, i.e. a bug in our tests
+2. Bugs in our testing infrastructure
+3. Bugs in Citus itself
+
+All of these impact the happiness and productivity of our developers, because we
+have to rerun tests to make them pass. But apart from developer happiness and
+productivity, 3 also impacts our users, and by ignoring flaky tests we can miss
+problems that our users could run into. Thus reducing the effectiveness of our
```
Comment (naisila) [inline src/test/regress/flaky_tests.md:14]:
```suggestion
problems that our users could run into. This reduces the effectiveness of our
```

File: src/test/regress/flaky_tests.md [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+## Easy fixes
+
+The following types of issues all fall within the category 1: bugs in our tests.
+
+### Expected records but different order
+
+**Issue**: A query returns the right result, but they are in a different order
+than expected by the output.
+
+**Fix**: Add an extra column to the ORDER BY clause to the query to make the
```
Comment (naisila) [inline src/test/regress/flaky_tests.md:50]:
```suggestion
**Fix**: Add an extra column to the ORDER BY clause of the query to make the
```

File: src/test/regress/flaky_tests.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
++NOTICE:  issuing SET citus.enable_ddl_propagation TO 'off'
++DETAIL:  on server postgres@localhost:57638 connectionId: 2
+ NOTICE:  issuing ANALYZE
+ DETAIL:  on server postgres@localhost:57638 connectionId: 2
+```
+
+**Example fix of issue 2**
+```diff
+ SET citus.log_remote_commands TO ON;
++SET citus.grep_remote_commands = '%ANALYZE%';
```
Comment (naisila) [inline src/test/regress/flaky_tests.md:137]: Cool, I didn't know we have this GUC.

**Review by naisila — APPROVED** [review]
Thanks for introducing this file.

some grammar fix suggestions from my part.

File: src/test/regress/flaky_tests.md (hunk 13 above)
Comment (JelteF) [reply src/test/regress/flaky_tests.md:138]: Nice to hear, people not knowing about some of these kind of tricks are exactly why I'm writing this document.

File: src/test/regress/flaky_tests.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+ test: multi_partitioning_utils replicated_partitioned_table
+-test: multi_partitioning partitioning_issue_3970
++test: multi_partitioning
++test: partitioning_issue_3970
+ test: drop_partitioned_table
+```
+
+### Use a fixed number of connections
+
+Citus its adaptive executor sometimes opens extra connections to do stuff in
```
[the remark this answers is not in the document]
Comment (JelteF) [reply src/test/regress/flaky_tests.md:239]: The contraction would actually be, but personally I don't think that looks/reads very nice so I'll keep it as is:
```suggestion
Citus' adaptive executor sometimes opens extra connections to do stuff in
```

[the remark this answers is not in the document]
Comment (naisila) [reply src/test/regress/flaky_tests.md:237]: `Citus's` is not a contraction of `Citus its`. I believe `Citus its` doesn't exist. If anything, `Citus's adaptive executor` can be called the contraction of `The adaptive executor of Citus`. So, the following suggestion is also grammatically correct:
```suggestion
The adaptive executor of Citus sometimes opens extra connections to do stuff in
```

Note: Citus' is also not grammatically correct. We use just the apostrophe, without the "s", when showing possession of plural nouns, not when nouns in general end with s. My point is, `Citus's` is correct, `Citus'` is not correct.

**Comment by onderkalaci** [comment]
even this PR is failing due to flaky tests :/

**Pull request merged**